src/geo/transform.js

```diff
@@ -1957,14 +1957,20 @@ class Transform {
     // Checks the four corners of the frustum to see if they lie in the map's quad.
     //
     isHorizonVisible(): boolean {
-        // we consider the horizon as visible if the angle between
-        // a the top plane of the frustum and the map plane is smaller than this threshold.
-        const horizonAngleEpsilon = 2;
-        if (this.pitch + radToDeg(this.fovAboveCenter) > (90 - horizonAngleEpsilon)) {
+        if (this.projection.name === 'mercator') {
+            // we consider the horizon as visible if the angle between
+            // a the top plane of the frustum and the map plane is smaller than this threshold.
+            const horizonAngleEpsilon = 2;
+            if (this.pitch + radToDeg(this.fovAboveCenter) > (90 - horizonAngleEpsilon)) {
+                return true;
+            }
+
+            return this.isCornerOffEdge(new Point(0, 0), new Point(this.width, this.height));
+        } else {
+            // complex shape of non-mercator maps means we cannot perform a cheap culling test
+            // so we deoptimize instead and always render the skybox.
```

Comment:
I think the math in this method and the first part of isCornerOffEdge is correct for any projection since it deals with the infinite plane, not the map surface. I think a `if (this.projection.name === 'mercator') return false` could be added here instead: https://github.com/mapbox/mapbox-gl-js/blob/28430749bd21212156151a379ff06b1336bbfa3b/src/geo/transform.js#L1922-L1924

Reply:
2cbd8b80ba0a7a507ffd1102b223c29ae9937b92